### PR TITLE
Add Subspaces and chats to an existing Space

### PR DESCRIPTION
**Pull request by kumarpalsinh25**
With this PR, the below features are added.

- Add subspaces to an existing Space
- Add sub-chats to an existing space
- Add recommended-space to an existing space

fixes: #1080

### CODE REVIEW

File: app/lib/features/space/pages/chats_page.dart
Comment (gnunicorn) [inline app/lib/features/space/pages/chats_page.dart:59]: what is this `extra: 1` for?

File: app/lib/features/space/pages/chats_page.dart
Comment (gnunicorn) [inline app/lib/features/space/pages/chats_page.dart:77]:
```suggestion
                              Text('Link existing Chat'),
```

File: app/lib/features/space/pages/chats_page.dart
Comment (gnunicorn) [inline app/lib/features/space/pages/chats_page.dart:73]: pretty sure we don't need both...

```suggestion
                            pathParameters: {'spaceId': spaceIdOrAlias},
```

File: app/lib/features/space/pages/related_spaces_page.dart
Comment (gnunicorn) [inline app/lib/features/space/pages/related_spaces_page.dart:81]:
```suggestion
                          pathParameters: {'spaceId': spaceIdOrAlias},
```

File: app/lib/features/space/pages/related_spaces_page.dart
Comment (gnunicorn) [inline app/lib/features/space/pages/related_spaces_page.dart:85]:
```suggestion
                            Text('Link existing Space'),
```

File: app/lib/features/space/sheets/link_chat_sheet.dart
Comment (gnunicorn) [inline app/lib/features/space/sheets/link_chat_sheet.dart:57]: use [`whenData`](https://pub.dev/documentation/hooks_riverpod/latest/hooks_riverpod/AsyncValueX/whenData.html) when only handling the data case?
```suggestion
    sp.whenData( (space) {
        subChatsIds.clear();
        for (int i = 0; i < space.knownChats.length; i++) {
          subChatsIds.add(space.knownChats[i].getRoomId().toString());
        }
      },);
```

File: app/lib/features/space/sheets/link_chat_sheet.dart
Comment (gnunicorn) [inline app/lib/features/space/sheets/link_chat_sheet.dart:69]:
```suggestion
      header: 'Link as Spacechat',
```

File: app/lib/features/space/sheets/link_chat_sheet.dart
Comment (gnunicorn) [inline app/lib/features/space/sheets/link_chat_sheet.dart:98]: It's not optional now, is it ;) .

```suggestion
        emptyText: '(none selected)`,
```

File: app/lib/features/space/sheets/link_chat_sheet.dart
Comment (gnunicorn) [inline app/lib/features/space/sheets/link_chat_sheet.dart:111]: I feel like this matching should already be done in the provider ...

File: app/lib/features/space/sheets/link_chat_sheet.dart
Comment (gnunicorn) [inline app/lib/features/space/sheets/link_chat_sheet.dart:140]: same as above.

File: app/lib/features/space/sheets/link_chat_sheet.dart
Comment (gnunicorn) [inline app/lib/features/space/sheets/link_chat_sheet.dart:161]: Similarly I'd expect all this in the same providers as well. Like as little of this as possible should be happening in the widget and rather in the providers...

File: app/lib/features/space/sheets/link_chat_sheet.dart
Comment (gnunicorn) [inline app/lib/features/space/sheets/link_chat_sheet.dart:202]: Probably was to show that error.

File: app/lib/features/space/sheets/link_space_sheet.dart
Comment (gnunicorn) [inline app/lib/features/space/sheets/link_space_sheet.dart:1]: Isn't this largely the same? Can't we make that a generic Widget then that we pass in some specific information?

The comments from the other file, still apply though.

File: app/lib/router/router.dart
Comment (gnunicorn) [inline app/lib/router/router.dart:383]: We have it in the path...

```suggestion
            parentSpaceId: state.pathParameters['spaceId'],
```

File: app/lib/router/router.dart
Comment (gnunicorn) [inline app/lib/router/router.dart:407]:
```suggestion
            parentSpaceId: state.pathParameters['spaceId'],
```

**Review by gnunicorn — CHANGES_REQUESTED** [review]
Looking good. But I wonder if we can't reuse the widget ...

Other than that just some minor style, code structuring and provider-usage things :) .

File: .changes/1106-add-subspace-and-chats.md
Comment (gtalha07) [inline .changes/1106-add-subspace-and-chats.md:1]: Minor typo ;)

File: app/lib/common/providers/space_providers.dart
Comment (gtalha07) [inline app/lib/common/providers/space_providers.dart:186]: would want to keep it ```.autoDispose``` so it flushes out when not listened?

File: app/lib/common/providers/space_providers.dart
Comment (gtalha07) [inline app/lib/common/providers/space_providers.dart:188]: Same here. I think ```.autoDispose()``` would be suitable modifier.

File: app/lib/features/space/sheets/link_chat_sheet.dart
Comment (gtalha07) [inline app/lib/features/space/sheets/link_chat_sheet.dart:30]: lets move it down after ```super.initState()``` ;)

File: app/lib/features/space/sheets/link_space_sheet.dart
Comment (gtalha07) [inline app/lib/features/space/sheets/link_space_sheet.dart:28]: Same here. Let's move it down after ```super.initState()``` .

File: app/lib/features/space/sheets/link_space_sheet.dart
Comment (gtalha07) [inline app/lib/features/space/sheets/link_space_sheet.dart:47]: use ```.whenData()``` if only handling data case and ```maybeWhen``` when you don't want to show ```error``` ```loading``` cases.

File: app/lib/features/space/sheets/link_space_sheet.dart
Comment (gtalha07) [inline app/lib/features/space/sheets/link_space_sheet.dart:165]: use ```const SizedBox.shrink()``` for just showing empty states.

**Review by gtalha07 — COMMENTED** [review]
Looks fine in general, but there are some suggestions which needed to be addressed before merge!

File: app/lib/features/space/pages/chats_page.dart
Comment (kumarpalsinh25) [reply app/lib/features/space/pages/chats_page.dart:59]: Not really aware about this param.  May be @gtalha07 can give better idea here.

File: app/lib/features/space/sheets/link_chat_sheet.dart
Comment (kumarpalsinh25) [reply app/lib/features/space/sheets/link_chat_sheet.dart:111]: Sorry but I was not able to find matching filter on the Chat Providers. Can you help me with the name of similar matching provider?

File: app/lib/features/space/sheets/link_space_sheet.dart
Comment (kumarpalsinh25) [reply app/lib/features/space/sheets/link_space_sheet.dart:1]: Yes true, even I was thinking to create generic widget at the beginning of the this work but don't exactly remember why I left that thought and created separate classes.

Let me thing again on this and do the needful.

File: app/lib/common/providers/space_providers.dart
Comment (kumarpalsinh25) [reply app/lib/common/providers/space_providers.dart:186]: I have taken reference from `chatSearchValueProvider`[ (chat_providers.dart)](https://github.com/acterglobal/a3/blob/main/app/lib/features/chat/providers/chat_providers.dart).

But ya you are right. It should be` .autoDispose`. I will do the required changes here.

File: app/lib/features/space/sheets/link_space_sheet.dart
Comment (kumarpalsinh25) [reply app/lib/features/space/sheets/link_space_sheet.dart:63]: fetchKnownSubspacesData() is for fetching selected parentSpace's child list. 

We have provide option to change the selected parentSpace in this screen. So in that case we need to fetch newly selected parentSpace's child list every time selection change. So that's why I have putted it here.

Please help me with your thoughts here in this scenario.

File: app/lib/features/space/pages/chats_page.dart
Comment (gtalha07) [reply app/lib/features/space/pages/chats_page.dart:60]: its a index of pages for ```PageView.builder()```, used for setting initial page of chat create sheet.

File: app/lib/features/space/sheets/link_space_sheet.dart
Comment (gtalha07) [reply app/lib/features/space/sheets/link_space_sheet.dart]: why not just have it triggered and updated based on user action? like ```onPressed(), onTap()``` which is on selection...

Comment (kumarpalsinh25) [reply app/lib/features/space/sheets/link_space_sheet.dart]: Let me check what can be done here. Thanks for sharing thoughts :)

File: app/lib/features/space/sheets/link_chat_sheet.dart
Comment (gnunicorn) [reply app/lib/features/space/sheets/link_chat_sheet.dart]: just create a new provider...

File: app/lib/features/space/pages/chats_page.dart
Comment (gnunicorn) [reply app/lib/features/space/pages/chats_page.dart:60]: so it isn't needed...
```suggestion
```

File: app/lib/features/space/sheets/link_room_sheet.dart
Comment (gnunicorn) [inline app/lib/features/space/sheets/link_room_sheet.dart:24]: I'd say the if-else in the initState below shouldn't happen here, but instead this should already be typed inputs and the routing controller should determine those:


```suggestion
  final String parentSpaceId;
  final ChildRoomType childRoomType;
```

File: app/lib/common/utils/routes.dart
Comment (gnunicorn) [inline app/lib/common/utils/routes.dart:42]: I'd still make this three different routes, that also elevates that query param problem....

```suggestion
  linkSubspace('/:spaceId([!#][^/]+)/linkSubspace'),
  linkChat('/:spaceId([!#][^/]+)/linkChat'),
  linkRecommended('/:spaceId([!#][^/]+)/linkRecommended'),
```

File: app/lib/features/space/sheets/link_room_sheet.dart
Comment (gnunicorn) [inline app/lib/features/space/sheets/link_room_sheet.dart:77]: It's never optional.

```suggestion
    final notifier = ref.read(selectedSpaceIdProvider.notifier);
    notifier.state = widget.parentSpaceId;
```

File: app/lib/features/space/sheets/link_room_sheet.dart
Comment (gnunicorn) [inline app/lib/features/space/sheets/link_room_sheet.dart:48]: not necessary then.
```suggestion
```

File: app/lib/features/space/sheets/link_room_sheet.dart
Comment (gnunicorn) [inline app/lib/features/space/sheets/link_room_sheet.dart:38]:
```suggestion
```

File: app/lib/features/space/sheets/link_room_sheet.dart
Comment (gnunicorn) [inline app/lib/features/space/sheets/link_room_sheet.dart:70]: Let's simplify this.

```suggestion
  String get sheetHeader {
    //Set data based on the childRoomType
    switch (widget.childRoomType) {
    case ChildRoomType.chat: 
       return 'Link as Space-chat';
    case ChildRoomType.recommendedSpace:
      return 'Link Recommended-Space';
    default:
      return 'Link Sub-Space';
    }
  }
```

File: app/lib/features/space/sheets/link_room_sheet.dart
Comment (gnunicorn) [inline app/lib/features/space/sheets/link_room_sheet.dart:102]: Not necessary right now, but this is a perfectly fine additional provider ;) .

File: app/lib/features/space/sheets/link_room_sheet.dart
Comment (gnunicorn) [inline app/lib/features/space/sheets/link_room_sheet.dart:130]:
```suggestion
        if (widget.childRoomType == ChildRoomType.chat) {
```

File: app/lib/features/space/sheets/link_room_sheet.dart
Comment (gnunicorn) [inline app/lib/features/space/sheets/link_room_sheet.dart:132]:
```suggestion
        } else if (widget.childRoomType == ChildRoomType.space) {
```

File: app/lib/features/space/sheets/link_room_sheet.dart
Comment (gnunicorn) [inline app/lib/features/space/sheets/link_room_sheet.dart:173]: No other cases exist, do they?

```suggestion
    if (widget.childRoomType == ChildRoomType.chat) {
      return chatsList();
    } else {
      return spacesList();
    }
```

File: app/lib/features/space/sheets/link_room_sheet.dart
Comment (gnunicorn) [inline app/lib/features/space/sheets/link_room_sheet.dart:190]: I know @gtalha07 had some feedback on having function that return `Widget`s, especially if they use `ref.watch` internally... Maybe you want to share...

**Review by gnunicorn — APPROVED** [review]
I'd manage those routes a bit differently, and generally move more towards strong typing - that simplifies quite a few things...

File: app/lib/features/space/sheets/link_room_sheet.dart
Comment (kumarpalsinh25) [reply app/lib/features/space/sheets/link_room_sheet.dart]: @gnunicorn 
Not getting your feedback, can you elaborate more about this feedback?

Comment (gnunicorn) [reply app/lib/features/space/sheets/link_room_sheet.dart]: I'd leave this to @gtalha07 ....

Comment (gtalha07) [reply app/lib/features/space/sheets/link_room_sheet.dart]: > I know @gtalha07 had some feedback on having function that return `Widget`s, especially if they use `ref.watch` internally... Maybe you want to share...

Could you confirm whether this works without any issues? The ```ref.watch()``` inside widget ```chatList()```?

Comment (kumarpalsinh25) [reply app/lib/features/space/sheets/link_room_sheet.dart]: I believe that `ref.watch()` will work inside widget `chatList()` without any issue.

The main reason is that it is used in the `ConsumerStatefulWidget` class and based on the riverpod core concept we can use `ref.watch()` anywhere in class which extends `ConsumerStatefulWidget`.

`ref` is define as a state in the riverpod `ConsumerStatefulWidget` class. 

Reference:
Screenshot 2023-11-02 at 9 20 44 PM

@gnunicorn 
Still share your thoughts here if I missed or miss understand anything about riverpod concept.

Comment (kumarpalsinh25) [reply app/lib/features/space/sheets/link_room_sheet.dart]: @gtalha07 
Still I’m unsure about the `ref.watch()` with one time listening or continues listening in a function.

@gnunicorn please share your thoughts here if have any idea about this concept

Comment (kumarpalsinh25) [reply app/lib/features/space/sheets/link_room_sheet.dart]: I haven't faced any issue on feature implementation.

**Pull request merged**
